docs: resource pricing specs (#2132)
celestiaorg/celestia-app: specs/src/specs/resource_pricing.md

consumption. For example, vesting accounts use more bytes of state than a normal
account, so more gas is consumed each time a vesting account is read from or
updated.

## Gas Limit

The gas limit must be included in each transaction. If the transaction exceeds
this gas limit during the execution of the transaction, then the transaction
will fail.

> Note: When a transaction is submitted to the mempool, the transaction is not
> fully executed. This can lead to a transaction getting accepted by the mempool
> and eventually included in a block, yet failing because the transaction ends
> up exceeding the gas limit.

Fees are not currently refunded. While users can specify a gas price,
the total fee is then calculated by simply multiplying the gas limit by the gas
price. The entire fee is then deducted from the transaction no matter what.

## Fee market

By default, Celestia uses mempools that prioritize fees, however mempool usage
cannot be enforced at the protocol level.

## Estimating PFB cost

The roughest way to estimate the total gas consumed by a PFB is by adding up the following:

- The cost for the transaction itself (~60k gas)

> - The cost of the blob. (number of shares used * 4096 gas)

Author: see #2116

	String() string
}
```

We can see how this gas meter is used in practice by looking at the store.
Notice where gas is consumed each time we write or read, specifically a flat
cost for initiating the action followed by a prorated cost for the amount of
data read or written.

```go
// Implements KVStore.
func (gs *Store) Get(key []byte) (value []byte) {
	gs.gasMeter.ConsumeGas(gs.gasConfig.ReadCostFlat, types.GasReadCostFlatDesc)
	value = gs.parent.Get(key)

	// TODO overflow-safe math?
	gs.gasMeter.ConsumeGas(gs.gasConfig.ReadCostPerByte*types.Gas(len(key)), types.GasReadPerByteDesc)
	gs.gasMeter.ConsumeGas(gs.gasConfig.ReadCostPerByte*types.Gas(len(value)), types.GasReadPerByteDesc)

	return value
}

// Implements KVStore.
func (gs *Store) Set(key []byte, value []byte) {
	types.AssertValidKey(key)
	types.AssertValidValue(value)
	gs.gasMeter.ConsumeGas(gs.gasConfig.WriteCostFlat, types.GasWriteCostFlatDesc)
	// TODO overflow-safe math?
	gs.gasMeter.ConsumeGas(gs.gasConfig.WriteCostPerByte*types.Gas(len(key)), types.GasWritePerByteDesc)
	gs.gasMeter.ConsumeGas(gs.gasConfig.WriteCostPerByte*types.Gas(len(value)), types.GasWritePerByteDesc)
	gs.parent.Set(key, value)
}
```

> The configuration for the gas meter used by Celestia is as follows.

Reviewer: [Non-blocking]
Is there a rationale for the values used?

Author: yeah, there's no way to change them at the moment :laughing:

Reviewer: What about the values?  Did we decide on them somewhere?

Author: sorry, what I meant is that we didn't discuss them or decide on them anywhere since they've never been able to be changed. These values are currently used for every cosmos-sdk chain. 

That's not to say that we shouldn't! We've occasionally mentioned that we could change these to optimize for specific things, such as aggregating data being posted or allowing many PFBs

Reviewer: Aah, thanks a lot

	}
}
```

Since gas is consumed in this fashion and many of the cosmos-sdk transactions
are composable, any given transaction can have a large window of possible gas
consumption. For example, vesting accounts use more bytes of state than a normal
account, so more gas is consumed each time a vesting account is read from or
updated.

## Parameters

There are three parameters that can be modified via governance to modify gas
usage.

| Parameter       | Default Value | Description                             |
|-----------------|---------------|-----------------------------------------|
| consensus/max_gas     | -1     | The maximum gas allowed in a block. Default of -1 means this value is not capped.             |
| auth/tx_size_cost_per_byte | 10 | Gas used per each byte used by the transaction. |
| auth/sig_verify_cost_secp256k1 | 1000 | Gas used per verifying a secp256k1 signature |
| blob/gas_per_blob_byte | 8 | Gas used per byte used by blob. Note that this value is applied to all encoding overhead, meaning things like the padding of the remaining share and namespace. See PFB gas estimation section for more details. |

> 

Reviewer: ```suggestion
```

## Gas Limit

The gas limit must be included in each transaction. If the transaction exceeds
this gas limit during the execution of the transaction, then the transaction
will fail.

> Note: When a transaction is submitted to the mempool, the transaction is not
> fully executed. This can lead to a transaction getting accepted by the mempool
> and eventually included in a block, yet failing because the transaction ends
> up exceeding the gas limit.

Fees are not currently refunded. While users can specify a gas price, the total
fee is then calculated by simply multiplying the gas limit by the gas price. The
entire fee is then deducted from the transaction no matter what.

## Fee market

By default, Celestia consensus nodes will use mempools that prioritize fees,
however mempool usage cannot be enforced at the protocol level. There is
currently no enforced minimum fee, this value is set by each consensus node in
their `app.toml`. Transactions that do not exceed that given gas price will not

> be able to enter the that node's mempool, and thus they will also not be

Reviewer 2: ```suggestion
be able to enter that node's mempool, and thus they will also not be
```

gossiped by that node.

## Estimating PFB cost

Generally, the gas used by a PFB transaction involves a static "fixed cost" and
a dynamic cost based on the size of each blob involved in the transaction.

> Note: For a general use case of a normal account submitting a PFB, the static
> costs can be treated as such. However, due to the description above of how gas
> works in the cosmos-sdk this is not always the case. Notably, if we use a
> vesting account or the `feegrant` modules, then these static costs change.

The "fixed cost" is an approximation of the gas consumed by operations outside
the function `GasToConsume` (for example, signature verification, tx size, read
access to accounts), which has a default value of 65,000.

> Note: the first transaction sent by an account (sequence number == 0) has an
> additional one time gas cost of 10,000. If this is the case, this should be
> accounted for.

> Each blob in the PFB contributes to the total gas cost based on its size. The
> function `GasToConsume` calculates the total gas consumed by all the blobs
> involved in a PFB, where each blob's gas cost is computed by multiplying the
> blob size (in bytes) by the `gasPerByte` parameter, along with adding a static
> amount per blob.

Reviewer 2: > is computed by multiplying the blob size (in bytes) by the `gasPerByte` parameter

nit: I think gas estimation determines the number of shares that are needed to contain the blob size and then multiples the number of shares * the number of bytes per share * `gasPerByte` parameter

Ref: https://github.com/celestiaorg/celestia-app/blob/3fc2fed84de759d3b82d80deff5df7f7ad7665b5/x/blob/types/payforblob.go#L161-L168

I took a stab at rewriting but feel free to rephrase: 

```suggestion
Each blob in the PFB contributes to the total gas cost based on its size. The
function `GasToConsume` calculates the total gas consumed by all the blobs
involved in a PFB, where each blob's gas cost is computed by first determining
how many shares are needed to store the blob size. Then, it computes the product
of the number of shares, the number of bytes per share, and the `gasPerByte`
parameter. Finally, it adds a static amount per blob.
```

The gas cost per blob byte and gas cost per transaction byte are parameters that
could potentially be adjusted through the system's governance mechanisms. Hence,
actual costs may vary depending on the current settings of these parameters.

## Tracing Gas Consumption

This figure plots each instance of the gas meter being incremented as a colored

> dot over the execution lifecycle of a given transaction. The y-axis in units of

Reviewer 2: ```suggestion
dot over the execution lifecycle of a given transaction. The y-axis is units of
```

gas and the x-axis is cumulative gas consumption. The legend shows which color
indicates what the cause of the gas consumption was.

> 

Reviewer: [non-blocking]
Linking to the code used to create these diagrams would be helpful in the future if things changes or someone wants to verify.

### MsgSend

Here we can see the gas consumption trace of a common send transaction for
1`utia`

![MsgSend](./figures/gas_consumption/msg_send_trace.png)

### MsgCreateValidator

Here we examine a more complex transaction.

![MsgCreateValidator](./figures/gas_consumption/msg_create_validator_trace.png)

### PFB with One Single Share Blob

![MsgPayForBlobs Single
Share](./figures/gas_consumption/single_share_pfb_trace.png)

### PFB with Two Single Share Blobs

This PFB transaction contains two single share blobs. Notice the gas cost for
`pay for blob` is double what it is above due to two shares being used, and
there is also additional cost from `txSize` since the transaction itself is
larger to accomodate the second set of metadata in the PFB.

![MsgPayForBlobs with Two
Blobs](./figures/gas_consumption/pfb_with_two_single_share_blobs_trace.png)

### 100KiB Single Blob PFB

Here we can see how the cost of a PFB with a large blob is quickly dominated by

> the cost of the blob.

Reviewer 2: 👍
